File: scala/98.scala
object Solution {
    def isValidBST(root: TreeNode): Boolean = {
        def test(root: TreeNode, min:Long, max: Long): Boolean = {
          if (root == null) true
          else if (root.value <= min || root.value >= max) false
          else test(root.left, min, root.value) && test(root.right, root.value, max)
        }
        test(root, Long.MinValue, Long.MaxValue)
    }
}

File: leetcodeJava/98.java
package leetcodeJava;

class Solution {
    public boolean isValidBST(TreeNode root) {
        return test(root, -(1l<<40), 1l<<40);
    }
    
    private boolean test(TreeNode root, long min, long max) {
        if (root == null) return true;
        if (root.val >= max || root.val <= min) return false;
        return test(root.left, min, root.val) && test(root.right, root.val, max);
    }
}
